src/on_curve.rs

```diff
@@ -15,7 +21,16 @@ impl<P: TEModelParameters> OnCurve for GroupProjective<P> {
 
         let on_curve = (YY + P::COEFF_A * XX) == (ZZ + P::COEFF_D * TT);
         let on_segre_embedding = self.t * self.z == self.x * self.y;
+        let z_non_zero = self.z != P::BaseField::zero();
+        let mut r_bytes = [0u8; 32];
+        (*constants::R)
+            .serialize(&mut r_bytes[..])
+            .expect("serialization into array should be infallible");
+        let r = P::ScalarField::from_le_bytes_mod_order(&r_bytes);
+        let mut two_r_bigint = r.into_repr();
+        two_r_bigint.mul2();
+        let point_order_2r = self.mul(two_r_bigint) == GroupProjective::zero();
 
-        on_curve && on_segre_embedding
+        on_curve && on_segre_embedding && z_non_zero && point_order_2r
```

Comment:
this is a total nit, and doesn't really matter for a debug function anyways, but as a style point, i think it'd be nice to wrap all this up in a scope, like
```
let point_order_2r = {
    let mut r_bytes = ...
    ....
    ...
    self.mul(two_r_bigint) == GroupProjective::zero()
};
```

so that it's clear what is the meaningful variable and what is the misc busywork needed to build it.